vavr/src/main/java/io/vavr/collection/Traversable.java

```diff
@@ -191,47 +188,62 @@
+     * <a href="https://en.wikipedia.org/wiki/Kahan_summation_algorithm">Kahan summation algorithm</a>
+     * (also known as compensated summation), which has known limitations.
+     * <p>
+     * Vavr uses Neumaier's modification of the Kahan algorithm, which yields better results.
+     *
+     * <pre>{@code
+     * // = Optional(0.0) (wrong)
+     * j.u.s.DoubleStream.of(1.0, 10e100, 2.0, -10e100).average()
+     *
+     * // = Some(0.75) (correct)
+     * List.of(1.0, 10e100, 2.0, -10e100).average()
+     * }</pre>
+     *
      * @return {@code Some(average)} or {@code None}, if there are no elements
      * @throws UnsupportedOperationException if this elements are not numeric
      */
-    @SuppressWarnings({ "unchecked", "OptionalGetWithoutIsPresent" })
     default Option<Double> average() {
-        if (isEmpty()) {
-            return Option.none();
-        } else {
-            final Traversable<?> objects = isTraversableAgain() ? this : toStream();
-            final Object o = objects.head();
-            if (o instanceof Number) {
-                final Traversable<Number> numbers = (Traversable<Number>) objects;
-                final double d;
-                if (o instanceof Integer || o instanceof Long || o instanceof Byte || o instanceof BigInteger || o instanceof Short) {
-                    d = numbers.toJavaStream()
-                            .mapToLong(Number::longValue)
-                            .average()
-                            .getAsDouble();
+        try {
+            @SuppressWarnings("unchecked")
+            final double[] sum = ((Traversable<Number>) this).map(Number::doubleValue)
+                    .foldLeft(new double[4], (acc, d) -> { // acc = [sum, c, count, simpleSum]
+                        final double tmp = acc[0] + d;
+                        acc[1] += (Math.abs(acc[0]) >= Math.abs(d)) ? (acc[0] - tmp) + d : (d - tmp) + acc[0];
+                        acc[0] = tmp;
+                        acc[2]++;
+                        acc[3] += d;
+                        return acc;
+                    });
+            final double count = sum[2];
+            if (count == 0) {
+                return Option.none();
+            } else {
+                final double avg = (sum[0] + sum[1]) / count;
+                final double simpleAvg = sum[3] / count;
+                if (Double.isNaN(avg) && Double.isInfinite(simpleAvg)) {
+                    return Option.some(simpleAvg);
                 } else {
-                    d = numbers.toJavaStream()
-                            .mapToDouble(Number::doubleValue)
-                            .average()
-                            .getAsDouble();
+                    return Option.some(avg);
                 }
-                return Option.some(d);
-            } else {
-                throw new UnsupportedOperationException("not numeric");
             }
+        } catch (ClassCastException x) {
```

Comment:
You should actualy include `x.getMessage()` as this will actually contain a message saying what type it actually had and tried to convert to.

Reply:
Thx for the suggestion, II've also thought about doing it. I agree that it is a good thing to be specific in order to make it easy to find the error reason. I decided not to return the cause, because the cause it is an implementation detail. But you are right, if we have the information then let's do it.

There is still a (low) risk to break applications that use Vavr in the case we decide to change the implementation in future and the exception cause disappears. Users should not rely on the cause. Our API spec tells that the method throws an UnsupportedOperationException.